### PR TITLE
TreeDB Raft: emit local-owner production route proof

### LINKED ISSUE #3467: TreeDB Raft #3460 M1B: local-owner production routed-commit proof and counters
Parent: #3460
Graph layer: next Layer 0
Depends on: #3462, #3461
Dispatch role: local-owner production route proof and counter slice

## Goal

Make the local-owner production routed-commit path measurable with counters that cannot be confused with local preflight-only route evidence.

Done means the benchmark/harness can emit `production_route_evidence` only when the command path performs real routed commits through the production submit/apply path, with counters proving route attempts, local-owner hits, commit group, applied group, and direct-local-bypass rejection behavior.

## Why This Exists

M0B intentionally made `-route-mode production` fail closed and marked ring-mode evidence as `local_preflight` / `production_scale_eligible=false`. The next production proof should start with the smallest true routed-commit path: local-owner writes that pass through the same production route gates future remote redirects, forwarding, and fanout will use.

## Current Evidence

- #3461 / PR #3465: local preflight evidence is explicitly non-production and production evidence schema is reserved.
- #3462 / PR #3464: current unsupported and unknown route shapes fail closed before mutation.
- Parent #3460 north-star gates require real routed commits, not local-only counters.

## Scope

- [ ] Add failing/expectation tests first for production route evidence emission boundaries.
- [ ] Populate `production_route_evidence` only when the harness exercised real routed commits.
- [ ] Count route attempts, local-owner hits, unknown-owner rejects, direct-local-bypass rejects, commit group hits, applied group hits, and leader/token partition hits where available.
- [ ] Keep `route_evidence` as local preflight evidence and keep `production_scale_eligible=false` outside the true production path.
- [ ] Ensure `-route-mode production` remains fail-closed until the production path is actually wired.
- [ ] Report latency and allocation fields only with clear measurement boundaries.

## Non-Goals And Boundary

- [ ] No remote-owner forwarding implementation.
- [ ] No cross-group fanout/split implementation.
- [ ] No 1-group vs N-group scaleout claim from local-owner-only proof.
- [ ] No catalog/meta ownership changes.
- [ ] No weakening of M0 fail-closed behavior.

## Required Tests

- [ ] `cmd/mongo_gateway_bench`: production mode fails closed before wiring, then emits `production_route_evidence` only for real routed commits.
- [ ] `cmd/mongo_gateway_bench`: local ring mode still emits `route_evidence` with `evidence_scope=local_preflight` and `production_scale_eligible=false`.
- [ ] `TreeDB/internal/raftcluster` or harness tests prove the local-owner command path went through group-routed submit/apply rather than a direct local bypass.
- [ ] Nativewire/Mongo smoke or focused tests cover local-owner routed commit success if this PR touches those surfaces.

Baseline command:

```sh
GOWORK=off go test ./TreeDB/internal/raftcluster ./TreeDB/nativewire ./TreeDB/mongo_gateway ./cmd/mongo_gateway_bench -count=1
```

## Performance Evidence

If the PR enables a timed production benchmark mode, include exact command, SHA, hardware/context, dataset shape, measured boundary, writes/sec, p50/p95/p99 write latency, `B/op`, `allocs/op`, and route/commit/apply counters. Before/after evidence is required for any hot-path overhead claim. A material regression blocks mergeability unless minimized and explicitly accepted with evidence.

## Branch And PR Policy

- [ ] Use a topic branch from current `main`.
- [ ] PR body links #3460 and this issue.
- [ ] Tests are added or tightened before broad implementation.
- [ ] Latest-head CI is required; stale green checks do not count.
- [ ] Request Codex, Copilot, and CodeRabbit only after the PR is mature: coherent code, focused tests, benchmark rationale/evidence, current PR body, no known local blockers, and latest-head CI running or green.
- [ ] Resolve or explicitly reject review findings with rationale before merge.

## Completion Criteria

- [ ] Local preflight evidence and production routed-commit evidence remain separate in JSON and docs.
- [ ] `production_route_evidence` is emitted only for true production routed commits.
- [ ] Counters prove local-owner routed submit/apply rather than direct local mutation.
- [ ] Required tests and relevant benchmark/evidence commands are recorded.
- [ ] PR is merged with latest-head CI and AI/code review findings resolved.

### PR DESCRIPTION
**Pull request by snissn**
Part of #3460.
Closes #3467.

## Summary
- Enables `cmd/mongo_gateway_bench -route-mode production` for a narrow local-owner proof: `-route-groups 1` only, BSON documents, command-WAL relaxed profile, serial insert-only workload shape.
- Adds a bench-only production route proof harness that routes collection creation and each measured insert through the production cluster submitter boundary.
- Emits `production_route_evidence` only when routed submit, commit, and apply counters prove true local-owner routed commits.
- Keeps local ring `route_evidence` as `evidence_scope=local_preflight` / `production_scale_eligible=false`; no production scaleout claim is made.
- Hardens the proof boundary after review: command-WAL disable fails closed, reused non-BSON collections are rejected, non-missing collection open errors are not masked, catalog-version known state is cleared when DB state is unavailable, unsupported production benchmark phases are rejected at parse time, catalog-version context cancellation is propagated, the production smoke test uses a bounded run context, route-evidence text assertions do not depend on Go map ordering, empty token-batch route requests fail closed, and forced GC is kept outside the profiled production load phase.

## Scope boundaries
- No remote-owner forwarding.
- No cross-group fanout/split execution.
- No catalog/meta ownership changes.
- No 1-group vs N-group horizontal-scale claim. Multi-group production mode remains fail-closed with `route-groups=1` required.
- Production evidence is currently limited to `-batch-size 1`, `-insert-producers 1`, no reads/range reads/updates/deletes, no secondary/range/update-indexed options, and no concurrent phases.

## Latest head
- `fd490188e` after aligning the production route profile boundary so forced GC happens before the profiled load phase and before allocation baseline capture.

## Tests
- `env GOWORK=off GOCACHE=/mnt/fast4tb/tmp/gomap-3467-gocache GOMODCACHE=/mnt/fast4tb/tmp/gomap-3467-gomodcache GOPATH=/mnt/fast4tb/tmp/gomap-3467-gopath GOTMPDIR=/mnt/fast4tb/tmp TMPDIR=/mnt/fast4tb/tmp go test ./cmd/mongo_gateway_bench -run 'Test(TreeDBProductionRouteModeLocalOwnerRoutedCommitSmoke|WriteResultProductionRouteEvidenceJSONAndText|ParseConfigRouteModeRing)$' -count=1`
- `env GOWORK=off GOCACHE=/mnt/fast4tb/tmp/gomap-3467-gocache GOMODCACHE=/mnt/fast4tb/tmp/gomap-3467-gomodcache GOPATH=/mnt/fast4tb/tmp/gomap-3467-gopath GOTMPDIR=/mnt/fast4tb/tmp TMPDIR=/mnt/fast4tb/tmp go test ./TreeDB/internal/raftcluster ./TreeDB/nativewire ./TreeDB/mongo_gateway ./TreeDB/mongo_gateway/benchsupport ./cmd/mongo_gateway_bench -count=1`

## Production Route Evidence Smoke
Command:

```sh
env GOWORK=off GOCACHE=/mnt/fast4tb/tmp/gomap-3467-gocache GOMODCACHE=/mnt/fast4tb/tmp/gomap-3467-gomodcache GOPATH=/mnt/fast4tb/tmp/gomap-3467-gopath GOTMPDIR=/mnt/fast4tb/tmp TMPDIR=/mnt/fast4tb/tmp \
  go run ./cmd/mongo_gateway_bench \
  -target treedb \
  -route-mode production \
  -route-groups 1 \
  -route-partitions 4 \
  -documents 100 \
  -batch-size 1 \
  -insert-producers 1 \
  -reads 0 -range-reads 0 -updates 0 -deletes 0 \
  -secondary-indexes 0 \
  -prebuild-documents \
  -treedb-maintenance none \
  -format json
```

Key result on latest measured head before the parse/context/test/profile-boundary review commits; the runnable shape is unchanged except the smoke test now uses the same single-producer proof:

```json
{
  "route_mode": "production",
  "route_group_count": 1,
  "route_partition_count": 4,
  "production_route_evidence_status": "available",
  "production_route_evidence": {
    "evidence_scope": "production_routed_commit",
    "real_routed_commits": true,
    "route_attempts_total": 101,
    "route_local_owner_hits": 101,
    "route_unknown_owner_rejects": 1,
    "direct_local_bypass_rejects": 1,
    "route_group_hits": {"group-00": 101},
    "commit_group_hits": {"group-00": 101},
    "applied_group_hits": {"group-00": 101},
    "fanout_split_attempts": 0,
    "fanout_split_failures": 0,
    "writes_per_sec": 753.7891094057051,
    "b_per_op": 472872.96,
    "allocs_per_op": 6736.82
  }
}
```

## Benchmark / regression assessment
This PR is an instrumentation/proof slice, not a throughput optimization. The smoke run is intentionally small and serial (`insert-producers=1`) to prove route/commit/apply counters without claiming scaleout. The allocation and latency fields are reported with an explicit runtime-memstats/load-phase boundary so downstream performance work can compare against a real production-route evidence object; forced GC now runs before the profiled load phase.

## AI review disposition
- Copilot proof-boundary findings were fixed: production mode fails closed on explicit command-WAL disable, reused non-BSON collections are rejected, non-missing collection-open errors are propagated, and stale catalog-version known state is cleared.
- Codex unsupported-production-phases finding was fixed by rejecting all non-serial insert-only production benchmark shapes at parse time, with docs and parser/smoke tests updated.
- Copilot context propagation finding was fixed by using the caller context in `ServerCatalogVersion` and returning cancellation errors from `CurrentCatalogVersion`.
- CodeRabbit smoke-test timeout nit was fixed by adding a bounded run context around the production routed-commit smoke test.
- Copilot map-order text assertion finding was fixed by checking map prefix and individual key/value substrings instead of a full ordered map string.
- Copilot empty-token-batch finding was fixed by rejecting empty token-batch route requests in the production proof recorder.
- Codex profile-boundary finding was fixed by moving forced GC outside the profiled production load phase.

## Summary by CodeRabbit

* **New Features**
  * Added production route-mode support for the Mongo gateway benchmark, including routed-commit proof runs, route evidence output, and additional performance metrics.
  * Expanded reporting to show route-group and partition details plus production evidence status and counters.

* **Documentation**
  * Updated benchmark guidance for production route-mode with clearer setup requirements, supported run shape, and how to read the new evidence output.

* **Tests**
  * Added coverage for production-mode config validation, evidence generation, and rejection handling.

### CODE REVIEW

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@coderabbitai review

**Comment by snissn** [comment]
@copilot review

File: cmd/mongo_gateway_bench/main.go [hunk 1]
```diff
@@ -665,6 +669,15 @@ func parseConfig(args []string) (config, error) {
...
 	cfg.RouteMode = routeMode
+	if cfg.Target == "treedb" && cfg.RouteMode == routeModeProduction {
+		cfg.TreeDBCommandWAL = true
+		if !seenFlags["treedb-profile"] {
+			treeDBProfile = string(treedb.ProfileCommandWALRelaxed)
+		}
+		if !seenFlags["treedb-document-format"] {
+			treeDBDocumentFormat = string(collections.DocumentFormatBSON)
+		}
+	}
```
[the remark this answers is not in the document]
Comment (snissn) [reply cmd/mongo_gateway_bench/main.go:683]: Fixed in the current head. Production mode now fails closed when treedb-command-wal is explicitly false instead of silently overriding it.

File: cmd/mongo_gateway_bench/main.go [hunk 2]
```diff
@@ -3693,6 +3759,112 @@ func runRoutedRingLoadPhase(ctx context.Context, cfg config, coll *mongo.Collect
...
+func ensureProductionRouteCollection(ctx context.Context, cfg config, target *benchTarget, db *mongo.Database) error {
+	if target != nil && target.collections != nil {
+		if _, err := target.collections.OpenCollection(cfg.Database + "." + cfg.Collection); err == nil {
+			return nil
+		}
+	}
+	if db == nil {
+		return errors.New("route-mode production requires Mongo database handle")
+	}
+	return db.RunCommand(ctx, bson.D{{Key: "create", Value: cfg.Collection}}).Err()
```
[the remark this answers is not in the document]
Comment (snissn) [reply cmd/mongo_gateway_bench/main.go:3885]: Fixed in the current head. Existing production proof collections must match the required BSON format, and non-missing OpenCollection errors are propagated instead of being treated as create-needed.

File: TreeDB/mongo_gateway/benchsupport/production_route.go [hunk 3]
```diff
@@ -0,0 +1,566 @@
...
+		return 0
+	}
+	state := p.db.State()
+	if state == nil {
+		return 0
+	}
+	p.value.Store(state.CommitSeq)
+	p.known.Store(true)
+	return state.CommitSeq
+}
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/mongo_gateway/benchsupport/production_route.go:311]: Fixed in the current head. Refresh now clears known/value state when DB state is unavailable, with benchsupport regression coverage.

File: cmd/mongo_gateway_bench/main.go [hunk 4]
```diff
@@ -665,6 +669,15 @@ func parseConfig(args []string) (config, error) {
 		return config{}, err
 	}
 	cfg.RouteMode = routeMode
+	if cfg.Target == "treedb" && cfg.RouteMode == routeModeProduction {
```
[the remark this answers is not in the document]
Comment (snissn) [reply cmd/mongo_gateway_bench/main.go:672]: Fixed in bdda14d89 and retained in the current head. route-mode production now rejects unsupported benchmark phases at parse time and accepts only the serial insert-only local-owner proof shape, with docs and parser/smoke tests updated.

File: TreeDB/mongo_gateway/benchsupport/production_route.go [hunk 5]
```diff
@@ -0,0 +1,571 @@
...
+func (p *productionRouteProofCatalogVersion) CurrentCatalogVersion(context.Context) (uint64, bool, error) {
+	if p == nil {
+		return 0, false, nil
+	}
+	p.Refresh()
+	return p.value.Load(), p.known.Load(), nil
+}
+
+func (p *productionRouteProofCatalogVersion) ServerCatalogVersion(context.Context) (uint64, error) {
+	version, ok, _ := p.CurrentCatalogVersion(context.Background())
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/mongo_gateway/benchsupport/production_route.go]: Fixed in 557df3fa9. ServerCatalogVersion now passes through the caller context, CurrentCatalogVersion returns cancellation errors, and a regression test fails if the adapter falls back to context.Background.

File: cmd/mongo_gateway_bench/main_test.go [hunk 6]
```diff
@@ -1124,6 +1162,116 @@ func TestParseConfigRouteModeRing(t *testing.T) {
...
+		defer cancel()
+		if err := closeBenchTarget(cleanupCtx, target); err != nil {
+			t.Errorf("cleanup target: %v", err)
+		}
+	}()
+
+	result, err := runBenchmark(context.Background(), cfg, target, nil)
+	if err != nil {
+		t.Fatalf("run production route benchmark: %v", err)
+	}
```
[the remark this answers is not in the document]
Comment (snissn) [reply cmd/mongo_gateway_bench/main_test.go:1210]: Fixed in 4383b1935. The production routed-commit smoke test now wraps runBenchmark in a 30s context and cancels it after the run; focused and broad affected suites passed on the updated head.

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@copilot review

**Comment by snissn** [comment]
@coderabbitai review

File: cmd/mongo_gateway_bench/main_test.go [hunk 7]
```diff
@@ -3401,6 +3557,28 @@ func TestProductionRouteEvidenceJSONSchemaPlaceholder(t *testing.T) {
...
+		"production_route_evidence_status=available",
+		"production_route_evidence evidence_scope=production_routed_commit",
+		"real_routed_commits=true",
+		"local_owner_hits=7",
+		"direct_local_bypass_rejects=3",
+		"commit_group_hits=map[group-00:6 group-01:3]",
+		"writes_sec=1234.500000",
+		"b_per_op=456.00",
+		"allocs_per_op=7.00",
+	} {
```
[the remark this answers is not in the document]
Comment (snissn) [reply cmd/mongo_gateway_bench/main_test.go:3581]: Fixed in 97a821556. The text-output test now checks commit_group_hits=map[ plus individual group-00 and group-01 substrings, avoiding Go map iteration order. Focused output tests and the broad affected suite passed on the updated head.

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@copilot review

**Comment by snissn** [comment]
@coderabbitai review

File: TreeDB/mongo_gateway/benchsupport/production_route.go [hunk 8]
```diff
@@ -0,0 +1,581 @@
...
+	var target nativewire.ClusterRouteTarget
+	switch request.Shape {
+	case nativewire.ClusterRouteShapeToken:
+		if !request.TokenKnown {
+			return nativewire.ClusterRouteTarget{}, errors.New("missing token")
+		}
+		target = r.routeToken(request.Token)
+	case nativewire.ClusterRouteShapeTokenBatch:
+		target = r.routeTokenBatch(request.Tokens)
+	case nativewire.ClusterRouteShapeCollection:
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/mongo_gateway/benchsupport/production_route.go:371]: Fixed in 67505d365. The production proof recorder now rejects empty token-batch route requests with missing token(s), with a benchsupport regression test. Focused and broad affected suites passed on the current head.

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@copilot review

**Comment by snissn** [comment]
@coderabbitai review

File: cmd/mongo_gateway_bench/main.go [hunk 9]
```diff
@@ -3693,6 +3799,120 @@ func runRoutedRingLoadPhase(ctx context.Context, cfg config, coll *mongo.Collect
...
+	var beforeDisk diskSnapshot
+	haveBeforeDisk := false
+	if target.treedbDir != "" {
+		if snapshot, err := collectDiskSnapshot(target.treedbDir); err == nil {
+			beforeDisk = snapshot
+			haveBeforeDisk = true
+		}
+	}
+
+	runtime.GC()
```
[the remark this answers is not in the document]
Comment (snissn) [reply cmd/mongo_gateway_bench/main.go]: Fixed in fd490188e. The production route forced GC now runs before runTreeDBProfiledPhaseWithDrain starts the load profile, so the CPU profile, phase timer, and allocation baseline share the same post-GC load boundary.

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@copilot review

**Comment by snissn** [comment]
@coderabbitai review
